SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp

```diff
@@ -186,19 +167,97 @@ QueryConstraint InterestFactory::CreateSystemDefinedConstraints()
 	return SystemDefinedConstraints;
 }
 
-QueryConstraint InterestFactory::CreateUserDefinedConstraints()
+QueryConstraint InterestFactory::CreateCheckoutRadiusConstraints()
 {
-	return QueryConstraint{};
-}
+	// Checkout Radius constraints are defined by the NetCullDistanceSquared property on actors.
+	//   - Checkout radius is a RelativeCylinder constraint on the player controller.
+	//   - NetCullDistanceSquared on AActor is used to define the default checkout radius with no other constraints.
+	//   - NetCullDistanceSquared on other actor types is used to define additional constraints if needed.
+	//   - If a subtype defines a radius smaller than a parent type, then its requirements are already captured.
+	//   - If a subtype defines a radius larger than all parent types, then it needs an additional constraint.
+	//   - Other than the default from AActor, all radius constraints also include Component constraints to
+	//     capture specific types, including all derived types of that actor.
+	//
+	// For best performance, the AActor NetCullDistanceSquared should be set to a value that captures the
+	// majority of actors important to a client, and actors outside that distance should be included by
+	// increasing the NetCullDistanceSquared property on that actor (or a parent actor type to capture a
+	// hierarchy). This should result in a minimal set of constraints for the default query.
+
+	const AActor* DefaultActor = Cast<AActor>(AActor::StaticClass()->GetDefaultObject());
+	const float DefaultDistanceSquared = DefaultActor->NetCullDistanceSquared;
+
+	// Gather NetCullDistance settings, and add any larger than the default radius to a list for processing.
+	TMap<UClass*, float> ActorCullDistances;
+	for (TObjectIterator<UClass> It; It; ++It)
+	{
+		if (!It->IsChildOf<AActor>())
+		{
+			continue;
+		}
 
-QueryConstraint InterestFactory::CreateCheckoutRadiusConstraint()
-{
-	QueryConstraint CheckoutRadiusConstraint;
+		const AActor* IteratedDefaultActor = Cast<AActor>(It->GetDefaultObject());
+		if (IteratedDefaultActor->NetCullDistanceSquared > DefaultDistanceSquared)
+		{
+			ActorCullDistances.Add(*It, IteratedDefaultActor->NetCullDistanceSquared);
+		}
+	}
+
+	// Sort the map for the next iteration so that base classes are seen before derived classes. This lets us skip
+	// derived classes that have a smaller cull distance than a parent class.
+	ActorCullDistances.KeySort([](UClass& LHS, UClass& RHS) {
+		return
+			LHS.IsChildOf(&RHS) ? -1 :
+			RHS.IsChildOf(&LHS) ? 1 :
+			0;
+	});
+
+	// If an actor's cull distance is smaller than that of a parent class, there's no need to add interest for that actor.
+	// Can't do inline removal since the sorted order is only guaranteed when the map isn't changed.
+	TMap<UClass*, float> OptimizedCullDistances;
+	for (const auto& ActorCullDistance : ActorCullDistances)
+	{
+		bool bShouldAdd = true;
+		for (auto& OptimizedCullDistance : OptimizedCullDistances)
```

Comment:
use of auto discouraged in Unreal coding standard except under certain circumstances, not sure about GDK stance:
https://docs.unrealengine.com/en-US/Programming/Development/CodingStandard/index.html#the'auto'keyword

Reply:
We are ok with putting `auto` in places like a for loop.